experimental/packages/otlp-exporter-base/test/common/configuration/otlp-http-configuration.test.ts

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  mergeOtlpHttpConfigurationWithDefaults,
+  OtlpHttpConfiguration,
+} from '../../../src/configuration/otlp-http-configuration';
+import * as assert from 'assert';
+import { testSharedConfigBehavior } from './shared-configuration.test';
+
+describe('mergeOtlpHttpConfigurationWithDefaults', function () {
+  const testDefaults: OtlpHttpConfiguration = {
+    url: 'http://default.example.test',
+    timeoutMillis: 1,
+    compression: 'none',
+    concurrencyLimit: 2,
+    headers: { 'User-Agent': 'default-user-agent' },
+  };
+
+  describe('headers', function () {
+    it('merges headers instead of overriding', function () {
+      const config = mergeOtlpHttpConfigurationWithDefaults(
+        {
+          headers: { foo: 'user' },
+        },
+        {
+          headers: { foo: 'fallback', bar: 'fallback' },
+        },
+        testDefaults
+      );
+      assert.deepStrictEqual(config.headers, {
+        'User-Agent': 'default-user-agent',
+        foo: 'user',
+        bar: 'fallback',
+      });
+    });
+
+    it('does not override default (required) header', function () {
+      const config = mergeOtlpHttpConfigurationWithDefaults(
+        {
+          headers: { 'User-Agent': 'custom' },
+        },
+        {
+          headers: { 'User-Agent': 'custom-fallback' },
+        },
+        testDefaults
+      );
+      assert.deepStrictEqual(config.headers, {
+        'User-Agent': 'default-user-agent',
+      });
+    });
+
+    it('drops user-provided headers with undefined values', function () {
+      const config = mergeOtlpHttpConfigurationWithDefaults(
+        {
+          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+          // @ts-ignore simulating plain JavaScript usage, ignoring types
+          headers: { foo: 'foo-user-provided', bar: undefined, baz: null },
+        },
+        {},
+        testDefaults
+      );
+      assert.deepStrictEqual(config.headers, {
+        foo: 'foo-user-provided',
+        baz: 'null',
```

Comment:
`bar: undefined` does not appear in the result but `baz: null` does and it's stringified. Any reason for that?

Reply:
This is mostly to keep the same behavior of the current implementation, which also handles it this way :slightly_smiling_face: 

### `undefined`

`bar: undefined` is dropped as calls with `http` module will throw when there's an `undefined` header value:

```
> const http = require('http');
undefined
> http.get({host: 'example.com', headers: {'foo': undefined}})
Uncaught:
TypeError [ERR_HTTP_INVALID_HEADER_VALUE]: Invalid value "undefined" for header "foo"
    at __node_internal_captureLargerStackTrace (node:internal/errors:496:5)
    at new NodeError (node:internal/errors:405:5)
    at __node_internal_ (node:_http_outgoing:624:11)
    at ClientRequest.setHeader (node:_http_outgoing:651:3)
    at new ClientRequest (node:_http_client:291:14)
    at request (node:http:100:10)
    at Object.get (node:http:111:15) {
  code: 'ERR_HTTP_INVALID_HEADER_VALUE'
}

```
We've had users in the past that have ended up with `undefined` header values as they set a header to `process.env.MY_HEADER_ENV_VAR`, did not set that env var and then ended up with an exporter that would constantly fail.

### `null`

`null` is stringified as to ensure it matches the type for `Record<string, string>` for the headers. Plus it gets stringified to `null` anyway when it's sent via with the `http` module so we make it happen earlier to make the type a bit more simple to work with down the chain as we don't have to worry about null values. It's also the current behavior of the exporters, so I tried not to change it.

You can try sending it to https://webhook.site to see that it ends up the same:

```
> const https = require('https');
undefined
> https.get({host: 'webhook.site', headers: {'foo': null}, path: '/<your-webhook-site-path>'})
> https.get({host: 'webhook.site', headers: {'foo': 'null'}, path: '/<your-webhook-site-path>'})
```